src/clib/map.d

```diff
@@ -143,12 +366,595 @@ struct map(K, V, A: IAllocator!V = allocator!V) {
             }
             node = node.next_node(key);
         }
-        return V.init; // Return default value if key not found
+        return V.init;
+    }
+
+    @property bool empty() const @nogc nothrow {
+        return root is null || root.filled == 0;
+    }
+
+    @property size_t size() const @nogc nothrow {
+        return _size;
+    }
+
+    bool contains(K key) const @nogc nothrow {
+        if (empty) return false;
+        return find_node(root, key) !is null;
+    }
+
+    private const(Node)* find_node(const Node* node, K key) const @nogc nothrow {
+        if (node is null) return null;
+        foreach (i; 0..node.filled) {
+            if (node.pairs[i].key == key) {
+                return node;
+            } else if (node.pairs[i].key > key) {
+                return find_node(node.children[i], key);
+            }
+        }
+        return find_node(node.children[node.filled], key);
+    }
+
+    const(V) opIndex(K key) const @nogc nothrow {
+        const(Node)* node = root;
+        while (node !is null) {
+            foreach (i; 0..node.filled) {
+                if (node.pairs[i].key == key) {
+                    return node.pairs[i].val;
+                }
+            }
+            node = node.next_node(key);
+        }
+        return V.init;
+    }
+
+    ref V opIndex(K key) @nogc nothrow {
+        Node* node = root;
+        while (node !is null) {
+            foreach (i; 0..node.filled) {
+                if (node.pairs[i].key == key) {
+                    return node.pairs[i].val;
+                }
+            }
+            node = node.next_node(key);
+        }
+
+        // Create new entry if not found
+        insert(key, V.init);
+
+        // Find the newly inserted value and return a reference to it
+        node = root;
+        while (node !is null) {
+            foreach (i; 0..node.filled) {
+                if (node.pairs[i].key == key) {
+                    return node.pairs[i].val;
+                }
+            }
+            node = node.next_node(key);
+        }
+        assert(0, "Internal error: newly inserted key not found");
```

Comment:
I'd highly prefer no asserts/throws. Return null if something fails. clib in itself is more like a C library rather then C++